hnn_core/gui/gui.py

```diff
@@ -39,232 +60,361 @@
+                colormap=plot_config['spectrogram_cm'],
+                ax=ax, show=False)
+
+    elif 'dipole@' in plot_type:
+        if len(dpls_copied) > 0:
+            plot_dipole(dpls_copied,
+                        ax=ax,
+                        layer=plot_type.split("@")[1],
+                        average=True,
+                        show=False)
+
+    elif plot_type == 'current dipole':
+        if len(dpls_copied) > 0:
+            plot_dipole(dpls_copied, ax=ax, average=True, show=False)
+
+    elif plot_type == 'layer-specific dipole':
+        if len(dpls_copied) > 0:
+            ax.remove()
+            layers = ["L2", "L5", "agg"]
+            gridspec = ax.get_subplotspec()
+            gs01 = gridspec.subgridspec(3, 1)
+            ax_l2 = fig.add_subplot(gs01[0])
+            ax_l5 = fig.add_subplot(gs01[1])
+            ax_lagg = fig.add_subplot(gs01[2])
+            axes = [ax_l2, ax_l5, ax_lagg]
+            plot_dipole(dpls_copied, ax=axes,
+                        layer=layers, average=True)
+        else:
+            print("No dipole data")
+
+    elif plot_type == 'network':
+        # raise NotImplementedError("No 3D plot for single ax now")
+        if net_copied:
+            _fig = plt.figure()
+            _ax = _fig.add_subplot(111, projection='3d')
+            net_copied.plot_cells(ax=_ax)
+
+            io_buf = io.BytesIO()
+            _fig.savefig(io_buf, format='raw')
+            io_buf.seek(0)
+            img_arr = np.reshape(np.frombuffer(io_buf.getvalue(),
+                                               dtype=np.uint8),
+                                 newshape=(int(_fig.bbox.bounds[3]),
+                                           int(_fig.bbox.bounds[2]), -1))
+            io_buf.close()
+            ax.imshow(img_arr)
+
+
+def _init_fig_from_mosaic(analysis_config, vis_plot_tree, viz_layout,
+                          init_vis_plot_tree, scale=0.95):
+    dpi = analysis_config['gui_config']['dpi']
+    mosaic = vis_plot_tree['mosaic']
+    assert len(mosaic) > 0
+
+    if init_vis_plot_tree:
+        vis_plot_tree['plots'] = {}
+
+    try:
+        kwargs = eval(f'dict({vis_plot_tree["kwargs"]})')
+    except Exception:
+        kwargs = {}
+    figsize = (scale * ((int(viz_layout.width[:-2]) - 10) / dpi),
+               scale * ((int(viz_layout.height[:-2]) - 10) / dpi))
+    for idx in plt.get_fignums():
+        plt.figure(idx).clear()
+    try:
+        fig, axd = plt.subplot_mosaic(mosaic, figsize=figsize, dpi=dpi,
+                                      **kwargs)
+    except Exception:
+        fig, axd = plt.subplot_mosaic(mosaic, figsize=figsize, dpi=dpi)
+
+    fig.canvas.header_visible = False
+
+    return fig, axd
+
+
+def _plot_sim_data(simulation_data, vis_plot_tree, analysis_config,
+                   init_vis_plot_tree=False):
+    viz_window = analysis_config['viz_window']
+    viz_layout = analysis_config['viz_style']
+    viz_window.clear_output()
```

Comment:
there is only one `Output` widget for the entire viz window? IMO, there should be one output widget per figure and they should be saved, so you don't have to re-render each time

Reply:
For example in the spiking plot, rasterization will be slow. If you save the raster image, then it can be rendered instantaneously ...

Reply:
That may actually add unnecessary complexity, as we will need to maintain another dict of fig objects. The current rendering time is now that fast but is still in an acceptable range

Reply:
> For example in the spiking plot, rasterization will be slow. If you save the raster image, then it can be rendered instantaneously ...

I see

Reply:
It's surely doable, while requiring some re-design. Give me some time to think and let me organize some of the current code. I hope I can have it slimmed and submitted before Sunday. 

There's some potential simplification we can apply to the current GUI, for example, we use some property "shortcuts" to simplify the function call. We may need to apply this throughly and if we maintain a list of outputs and corresponding figs then we need to also integrate that into those shortcuts.